homeassistant/components/text/__init__.py

```diff
@@ -0,0 +1,194 @@
+"""Component to allow setting text as platforms."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from datetime import timedelta
+import logging
+import re
+from typing import Any, final
+
+import voluptuous as vol
+
+from homeassistant.backports.enum import StrEnum
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import MAX_LENGTH_STATE_STATE
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    ATTR_MAX,
+    ATTR_MIN,
+    ATTR_MODE,
+    ATTR_PATTERN,
+    ATTR_VALUE,
+    DOMAIN,
+    SERVICE_SET_VALUE,
+)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+_LOGGER = logging.getLogger(__name__)
+
+# mypy: disallow-any-generics
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Text entities."""
+    component = hass.data[DOMAIN] = EntityComponent[TextEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALUE,
+        {vol.Required(ATTR_VALUE): cv.string},
+        _async_set_value,
+    )
+
+    return True
+
+
+async def _async_set_value(entity: TextEntity, service_call: ServiceCall) -> None:
+    """Service call wrapper to set a new value."""
+    value = service_call.data[ATTR_VALUE]
+    if len(value) < entity.min:
+        raise ValueError(
+            f"Value {value} for {entity.name} is too short (minimum length {entity.min})"
+        )
+    if len(value) > entity.max:
+        raise ValueError(
+            f"Value {value} for {entity.name} is too long (maximum length {entity.max})"
+        )
+    if entity.pattern is not None and re.compile(entity.pattern).match(value) is None:
+        raise ValueError(
+            f"Value {value} for {entity.name} doesn't match pattern {entity.pattern}"
+        )
+    await entity.async_set_value(value)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class TextMode(StrEnum):
+    """Modes for text entities."""
```

Comment:
We have a lot more "modes" when I look at the `TextSelector`, should this be extended?

Reply:
Eventually yes but that would require frontend changes too right? If so maybe that should be in a separate PR. Also begs the question, should we have predefined regex patterns for some of these types?

Reply:
> Eventually yes but that would require frontend changes too right?

It does require that now already? As in, to support `password`, we cannot rely only on the backend?

Reply:
doesn't the text input component already support password?

Reply:
Yes, but it uses `<ha-textfield>` and mode directly passes in as the `type`:

![image](https://user-images.githubusercontent.com/195327/203161658-c3af4048-5021-430b-8299-1530ffe4ad1d.png)

Which is the same as the selector does as it seems. From the `ha-selector-text`:

![image](https://user-images.githubusercontent.com/195327/203161805-3c93d908-3744-4dac-910c-22daa03d53de.png)

Which is defined/coming from the MWC Textfield:

![CleanShot 2022-11-21 at 22 30 17](https://user-images.githubusercontent.com/195327/203162044-aeaad888-8ad8-455f-acf2-d4cb7b9a7143.png)

So... I think it can just be extended and it will work?

Reply:
instead of redefining it, should we just reuse the existing class? Maybe define it in `homeassistant.const` since it will presumably be used in mutiple places?

Reply:
> instead of redefining it, should we just reuse the existing class? Maybe define it in homeassistant.const since it will presumably be used in mutiple places?

The number of places is very limited at this point, so wouldn't go that path just yet.

> also should we exclude date, datetime-local and time since they will be supported by the other new platforms?

Agree, number, date, month, week time, datetime-local, and colorsare  maybe odd..

Hmmm ok that removes a lot. OK, convinced we should maybe just stick with text  & password for now 😬